File: leetcode-practice-java/src/main/java/io/hank/leetcode/practices/_001_Two_Sum.java
package io.hank.leetcode.practices;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

/**
 * Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
 * <p>
 * You may assume that each input would have exactly one solution, and you may not use the same element twice.
 * <p>
 * You can return the answer in any order.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * Input: nums = [2,7,11,15], target = 9
 * <p>
 * Output: [0,1]
 * <p>
 * Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
 * <p>
 * Example 2:
 * <p>
 * Input: nums = [3,2,4], target = 6
 * <p>
 * Output: [1,2]
 * <p>
 * Example 3:
 * <p>
 * Input: nums = [3,3], target = 6
 * <p>
 * Output: [0,1]
 * <p>
 * <p>
 * Constraints:
 * <p>
 * 2 <= nums.length <= 10^4
 * <p>
 * -10^9 <= nums[i] <= 10^9
 * <p>
 * -10^9 <= target <= 10^9
 * <p>
 * Only one valid answer exists.
 * <p>
 * <p>
 * Follow-up: Can you come up with an algorithm that is less than O(n^2) time complexity?
 */
public class _001_Two_Sum {
    public void execute() {
        int[] result1 = twoSum(new int[]{2, 7, 11, 15}, 9);
        System.out.println(Arrays.toString(result1));

        int[] result2 = twoSum(new int[]{2, 3, 4}, 6);
        System.out.println(Arrays.toString(result2));

        int[] result3 = twoSum(new int[]{3, 3}, 6);
        System.out.println(Arrays.toString(result3));
    }

    int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> numMap = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i];
            if (numMap.containsKey(complement)) {
                return new int[]{numMap.get(complement), i};                
            }
            numMap.put(nums[i], i);
        }

        return new int[]{}; // no matched
    }
}

File: leetcode-practice-kotlin/src/main/kotlin/io/hank/leetcode/kotlin/practices/_001_Two_Sum.kt
package io.hank.leetcode.kotlin.practices

import java.util.*

/**
 * Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
 *
 * You may assume that each input would have exactly one solution, and you may not use the same element twice.
 *
 * You can return the answer in any order.
 *
 * Example 1:
 *
 * Input: nums = [2,7,11,15], target = 9
 *
 * Output: [0,1]
 *
 * Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
 *
 * Example 2:
 *
 * Input: nums = [3,2,4], target = 6
 *
 * Output: [1,2]
 *
 * Example 3:
 *
 * Input: nums = [3,3], target = 6
 *
 * Output: [0,1]
 *
 * Constraints:
 *
 * 2 <= nums.length <= 10^4
 *
 * -10^9 <= nums[i] <= 10^9
 *
 * -10^9 <= target <= 10^9
 *
 * Only one valid answer exists.
 *
 * Follow-up: Can you come up with an algorithm that is less than O(n^2) time complexity?
 */
class _001_Two_Sum : LeetcodeProblemSolution() {
    private fun twoSum(nums: IntArray, target: Int): IntArray {
        val numMap = HashMap<Int, Int>()
        nums.forEachIndexed { index, element ->
            val complement = target - element
            numMap.takeIf { it.containsKey(complement) }?.let { map ->
                return intArrayOf(map[complement] ?: 0, index)
            }
            numMap[nums[index]] = index
        }
        return intArrayOf()
    }

    override fun execute() {
        super.execute()

        // Example 1
        val input1 = intArrayOf(2, 7, 11, 15)
        val target1 = 9
        println("Input: nums = ${input1.contentToString()}, target = $target1")
        val result1 = twoSum(input1, target1)
        println("Output: ${result1.contentToString()}")

        // Example 2
        val input2 = intArrayOf(2, 3, 4)
        val target2 = 6
        println("Input: nums = ${input2.contentToString()}, target = $target2")
        val result2 = twoSum(input2, target2)
        println("Output: ${result2.contentToString()}")

        // Example 1
        val input3 = intArrayOf(3, 3)
        val target3 = 6
        println("Input: nums = ${input3.contentToString()}, target = $target3")
        val result3 = twoSum(input3, target3)
        println("Output: ${result3.contentToString()}")
    }
}
